lnwallet/musig_session.go

```diff
@@ -0,0 +1,556 @@
+// nonce state to disk. Instead, we can use this to derive the nonce we need to
+// sign and broadcast our own commitment transaction.
+func WithLocalCounterNonce(targetHeight uint64,
+	shaGen shachain.Producer) MusigSessionOpt {
+
+	return func(opt *musigSessionOpts) {
+		nextPreimage, _ := shaGen.AtIndex(targetHeight)
+
+		opt.customRand = bytes.NewBuffer(nextPreimage[:])
+	}
+}
+
+// VerifyCommitSig attempts to verify the passed partial signature against the
+// passed commitment transaction. A keyspend sighash is assumed to generate the
+// signed message. As we never re-use nonces, a new verification nonce (our
+// relative local nonce) returned to transmit to the remote party, which allows
+// them to generate another signature.
+func (m *MusigSession) VerifyCommitSig(commitTx *wire.MsgTx,
+	sig *lnwire.PartialSigWithNonce,
+	musigOpts ...MusigSessionOpt) (*musig2.Nonces, error) {
+
+	opts := defaultMusigSessionOpts()
+	for _, optFunc := range musigOpts {
+		optFunc(opts)
+	}
+
+	// Before we can verify the signature, we'll need to finalize the
+	// session by binding the remote party's provided signing nonce.
+	if err := m.FinalizeSession(musig2.Nonces{
+		PubNonce: sig.Nonce,
+	}); err != nil {
+		return nil, err
+	}
+
+	// When we verify a commitment signature, we always assume that we're
+	// verifying a signature on our local commitment. Therefore, we'll use:
+	// their remote nonce, and also public key.
+	partialSig := NewMusigPartialSig(
+		&musig2.PartialSignature{S: &sig.Sig},
+		m.nonces.SigningNonce.PubNonce, m.combinedNonce, m.signerKeys,
+	)
+
+	// With the partial sig loaded with the proper context, we'll now
+	// generate the sighash that the remote party should have signed.
+	sigHash, err := taprootKeyspendSighash(
+		commitTx, m.inputTxOut.PkScript, m.inputTxOut.Value,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	walletLog.Infof("Verifying new musig2 sig for session=%x, nonce=%s",
+		m.session.SessionID[:], m.nonces.String())
+
+	if !partialSig.Verify(sigHash, m.remoteKey.PubKey) {
+		return nil, fmt.Errorf("invalid partial commit sig")
+	}
+
+	nonceOpts := []musig2.NonceGenOption{
+		musig2.WithPublicKey(m.localKey.PubKey),
+	}
+	if opts.customRand != nil {
+		nonceOpts = append(
+			nonceOpts, musig2.WithCustomRand(opts.customRand),
+		)
+	}
+
+	// At this point, we know that their signature is valid, so we'll
+	// generate another verification nonce for them, so they can generate a
+	// new state transition.
+	nextVerificationNonce, err := musig2.GenNonces(nonceOpts...)
+	if err != nil {
+		return nil, fmt.Errorf("unable to gen new nonce: %w", err)
+	}
+
+	return nextVerificationNonce, nil
+}
+
+// CombineSigs combines the passed partial signatures into a valid schnorr
+// signature.
+func (m *MusigSession) CombineSigs(sigs ...*musig2.PartialSignature,
+) (*schnorr.Signature, error) {
+
+	defer func() {
+		err := m.signer.MuSig2Cleanup(m.session.SessionID)
+		if err != nil {
```

Comment:
I think `MuSig2CombineSig` will clean up the session so this will always return an error because the session is not found. Maybe we shouldn't do the auto cleanup in `MuSig2CombineSig`, or mention it in the method name or docs since it's error-prone.

Reply:
Ah yeah good point, removed since it isn't necessary. 

Initially I wanted to try to clean up the session as soon as possible, but realized that you need the session state present in order to do combine (so you can't clean up, then try to combine later the state won't be found). What I think we should do is add a `defer CleanUp()` in the link itself, so this way whenever we hang up the connection we remove the active session. We then also want to do this in the state machine after a round trip is complete. 

I need to revisit the force close code in the final part of this PR, I think it ends up re-creating the session so it can properly force close.